packages/phone-number-privacy/src/index.ts

```diff
@@ -1,21 +1,148 @@
+    return BLINDBLS.computePRF(privateKey, new Buffer(queryPhoneNumber))
   } catch (e) {
-    console.error('Failed to compute BLS salt', e)
-    response.status(500).send('Failed to compute BLS salt')
+    console.error('Failed to compute salt', e)
+    response.status(500).send('Failed to compute salt')
   }
-})
+}
+
+/*
+ * Returns how many queries the account can make based on the
+ * calculated query quota and the number of queries already performed.
+ */
+async function getRemainingQueryCount(account: string, phoneNumber: string) {
+  const queryQuota = await getQueryQuota(account, phoneNumber)
+  const performedQueryCount = await getPerformedQueryCount(account)
+  return queryQuota - performedQueryCount
+}
+
+/*
+ * Calculates how many queries the caller has unlocked based on the algorithm
+ * unverifiedQueryCount + verifiedQueryCount + (queryPerTransaction * transactionCount)
+ */
+async function getQueryQuota(account: string, phoneNumber: string) {
+  let queryQuota: number = config.salt.unverifiedQueryCount
+  if (await isVerified(account, phoneNumber)) {
+    queryQuota += config.salt.verifiedQueryCount
+    const transactionCount = await getTransactionCountFromAccount(account)
+    queryQuota += config.salt.queryPerTransaction * transactionCount
+  }
+  return queryQuota
+}
+
+async function isVerified(account: string, phoneNumber: string): Promise<boolean> {
+  const attestationsWrapper: AttestationsWrapper = await contractKit.contracts.getAttestations()
+  const attestationStats = await attestationsWrapper.getAttestationStat(phoneNumber, account)
+  const numAttestationsCompleted = attestationStats.completed
+  const numAttestationsRemaining = NUM_ATTESTATIONS_REQUIRED - numAttestationsCompleted
+  return numAttestationsRemaining <= 0
+}
+
+async function getTransactionCountFromAccount(account: string): Promise<number> {
+  // TODO [amyslawson] consider adding retry
+  const transactionUrl = `${config.blockchain.blockscout}/api?module=account&action=tokentx&address=${account}`
+  const resp = await fetch(transactionUrl)
+  const jsonResp = await resp.json()
+  if (jsonResp == null || jsonResp.result == null) {
+    console.error(`Failed to get valid response for ${transactionUrl}`)
+    return 0
+  }
+  return jsonResp.result.filter((transaction: any) => transaction.from === account).length
+}
+
+/*
+ * Returns how many queries the account has already performed.
+ */
+async function getPerformedQueryCount(account: string): Promise<number> {
+  return knex(ACCOUNTS_TABLE)
+    .where(ACCOUNTS_COLUMNS.address, account)
+    .select(ACCOUNTS_COLUMNS.numLookups)
+    .first()
+    .then((object) => {
+      return object === undefined ? 0 : object[ACCOUNTS_COLUMNS.numLookups]
+    })
+    .catch((reason) => {
+      // TODO [amyslawson] think of failure case here
+      console.error(reason)
+      return 0
+    })
+}
+
+/*
+ * Increments query count in database.
+ */
+async function incrementQueryCount(account: string) {
```

Comment:
The query and increment logic should be atomic. For example, if I hit two servers simultaneously, both query my count as 32 and increment my count by 1, then my count will reach 34, which breaks the rules.

Some ideas to fix this:
1. Use serializable transactions. This will ensure that only one thread can read and modify the tx at a time. The second update will fail which is desired in our scenario.
See "SERIALIZABLE transactions" [here](https://www.2ndquadrant.com/en/blog/postgresql-anti-patterns-read-modify-write-cycles/).
2. Pass in the read value ('expectedLookups') into incrementQueryCount and fail if it's different than expected.
```
knex(ACCOUNTS_TABLE)
      .where(ACCOUNTS_COLUMNS.address, account)
      .where(ACCOUNTS_COLUMNS.numLookups, expectedLookups)
      .increment(ACCOUNTS_COLUMNS.numLookups, 1)
```
See "Optimistic concurrency controls" in the linked article.
3. Increment first, then read
This has the negative side-effect of potentially blocking a user too soon. However, if a user is spamming us perhaps they deserve to be blocked 😝

Reply:
this is a good point.
1. it seems we can pass this setting in via a raw query.  it says it lasts the current session (which I would think it just the full firebase request that creates the connection to the DB)
2. the only complication here with knex is that i'm already doing an update OR create, so if the update fails, its assuming its because the entry doesn't exist.  if I add this additional condition, it will also try to perform the create (which would also fail since it exists, but its just wasted requests)
3. ya -- we could do this and would only be an issue if there's an error later that prevents the user from getting the salt (we've essentially wasted their query)

i'm not sure how to test -- particularly the first one would be hard to tell if its been implemented correctly

Reply:
I added a sleep to simulate a race condition.  With adding
`  await knex.raw('BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;')`

it too also will fail on the update and then try the insert.  there really is no way to tell the difference between a failed update due to an entry nonexistence vs rejection to race condition.  the db returns just "0" to represent 0 rows updated rather than throw a parsable error :/

oddly the second request to "insert" then hangs and requires adding a timeout.

lets discuss in our next meeting